### PR TITLE
Issue #3555: refactoring: CodeSelectorPModel do double copy of collection in c-tor

### PR DESCRIPTION
**Pull request by Kietzmann**
Issue #3555 
What was made:
Replaced creating temporary list and then creating unmodifiable view with `Collections.unmodifiableList` with assigning `line2position` field to new ArrayList and then adding all elements from origin list to field with `addAll()` method. All tests passed correctly/ During manual test case I checked that modifications of element's in original collection will not modify elements in copy as was supposed.

### CODE REVIEW

**Comment by romani** [comment]
@Kietzmann , before sending PR, please run "mvn verify" on your local it will economize you bunch of time.

**Comment by Kietzmann** [comment]
@romani 
This time I reluctanted to perform `mvn verify` again after minor checks)

**Comment by rnveach** [comment]
@Kietzmann You made changes to `CodeSelectorPresentation`. In my post, I mentioned 2 other areas where we are copying the list and making it unmodifiable. https://github.com/checkstyle/checkstyle/issues/3555#issuecomment-261057159
Should the other areas be changed too?

**Comment by Kietzmann** [comment]
@rnveach 
I changed in 2 other areas you'd mentioned. I also find usage of `Collections.unmodifiableList` in other classes
[TestRootModuleChecker](https://github.com/checkstyle/checkstyle/blob/48f9e33c3f73f50051e8618ae83aa9444bae9784/src/test/java/com/puppycrawl/tools/checkstyle/TestRootModuleChecker.java#L73)
[VisibilityModifierCheck 1](https://github.com/checkstyle/checkstyle/blob/55aa8cedc60d4faa776d7a36f8ae2d0bed0ada45/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/VisibilityModifierCheck.java#L249)
[VisibilityModifierCheck 2](https://github.com/checkstyle/checkstyle/blob/55aa8cedc60d4faa776d7a36f8ae2d0bed0ada45/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/VisibilityModifierCheck.java#L274)
[AbstractHeaderCheck](https://github.com/checkstyle/checkstyle/blob/ce21086e087661553f3a774c38362327ee88996a/src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java#L74)
[JavadocTag 1](https://github.com/checkstyle/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTags.java#L44)
[JavadocTag 2](https://github.com/checkstyle/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTags.java#L46)
[JavadocTag 3](https://github.com/checkstyle/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTags.java#L46)
[JavadocTag 4](https://github.com/checkstyle/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTags.java#L46)

Should I change them as previous or leave as it was?
Thanks

**Comment by rnveach** [comment]
> Should I change them as previous or leave as it was?

The issue isn't with `unmodifiableList`.
The issue is taking an existing list, and copying it to a new list multiple times over.
The areas I mentioned in my post was the same array being duplicated over and over again. If we make the list modifiable once, then the sub-classes don't need to do the same work again, right?

**Comment by Kietzmann** [comment]
@rnveach 
As I've seen in those files you've mentioned we just using final collection without any modification of it's content. Yes, we don't need to make new unmodList from unmodList, we just can assign ref to it.

**Comment by Kietzmann** [comment]
@rnveach 
So what should I do next? Seems that issue was is understanding of usage of unmodifiableList in those cases, or I should make some modifications to current PR?

**Comment by MEZk** [comment]
@Kietzmann 
> As I've seen in those files you've mentioned we just using final collection without any modification of its content. Yes, we don't need to make new unmodList from unmodList, we just can assign a ref to it.

One again, the issue is not with the unmodifiable list. @rnveach pointed that we create the copy of the original list multiple times:

> I think the code at MainFrameModel.java#L177 is valid. All the other array copies and 'unmodifiable' may be just extra protection to make sure we are not passing around a list that can be modified. One class won't know what its children are doing with it.

**Comment by Kietzmann** [comment]
@MEZk 
So what's the problem? Seems that my modifications is unnecessary in that way

**Comment by rnveach** [comment]
We need to limit creating new lists, and constantly making new list unmodifiable just for sub-classes.

https://github.com/checkstyle/checkstyle/blob/d49732c325b15cb6d7d34f7f838bd51ced2b8e3d/src/main/java/com/puppycrawl/tools/checkstyle/gui/MainFrameModel.java#L210 
I think we should change this to create a new array on each file open, and store it into the field as an unmodifiable list.
Every other retrieval of the list should then just return/get the list as is. No more need to create new ones.

@MEZk @romani This sound good?

**Comment by MEZk** [comment]
@rnveach 
> I think we should change this to create a new array on each file open,

Why do we need the array? It will be transformed to the list. We know the size of the [sourceLines](https://github.com/MEZk/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/gui/MainFrameModel.java#L207-L207), so we can create the array list of that size + 1 (specify the size of the inner array of ArrayList). I think that in this particular case the cost of the array transformation to the array list is comparable with the new array list creation.

Thus, my proposal is:
1) Create the new array list for each file opened;
2) Make the list unmodifiable as you suggested (the list will be the unmodifiable view of the source list);
3) Store the list as a field;
4) Give the unmodifiable list to all classes which invokes [getLinesToPosition](https://github.com/MEZk/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/gui/MainFrameModel.java#L175-L175).

**Comment by rnveach** [comment]
> Why do we need the array?

Yes, array list.

@Kietzmann Please make the additional change.

**Comment by romani** [comment]
@rnveach , @Kietzmann , sorry for delay in reply ...... I did review of this issue

so code is weird by its over protection for this collection only , it is not required.
Collection is created in MainFrameModel --> MainFrame --> TreeTable --> CodeSelector --> CodeSelectorPresentation
If we want to be protective and reasonable in same time, all we need is too make a copy and wrap by `unmodifiableList` on the moment when we return this collection from MainFrameModel  - it is the only object that make it and he has all right to be over protective, as its expose its inner state. All other classes do not need any copy or wrapping at all !! They will receive unmodifiable collection and it is enough, it they damage it - it is their problems, serious issue in logic. As we keep source of this collection protected we could reinitialize all other classes again if required.

If both of you agree, lets remove this copy-wrap madness in all classes except for MainFrameModel. We should put a javadoc comment on code in MainFrameModel to explain why copy is happening in this place for future engineers.

**Comment by rnveach** [comment]
> all we need is too make a copy and wrap by unmodifiableList on the moment when we return this collection from MainFrameModel

@romani Me and MEZk think we should do this when we create the list, not when we return it in getter. This way we are not making new lists every call to getter, but only when a file is opened.

> If both of you agree, lets remove this copy-wrap madness in all classes except for MainFrameModel.

I agree.

**Comment by Kietzmann** [comment]
@romani @rnveach 
Ok

**Comment by Kietzmann** [comment]
@rnveach 
But in that case list field is class can't be final, cause we reassigning it unmodifiable list after file is opened
@romani variant is more desirable

**Comment by romani** [comment]
> This way we are not making new lists every call to getter, but only when a new file is opened.

yes , we could it like this, lets do it but with lazy initialization of copy in getter.

But now we have IntelijIdea complains about "Assignment to Collection field this.lines2position from parameter 'lines2position' " that may be a reason why such code exists :)

**Comment by MEZk** [comment]
@romani 
Why do we need to follow IDEA Inspection if it is not appropriate in the current situation?

> To make all happy we should stay with 3 copies
in MainFrameModel

So, do you support here our idea to create a new list each time the file is opened?

> out of MainFrameModel in the way to destination

You mean in getter? Why?

> in destination (CodeSelectorPresentation)

Just to make IDEA Inspection happy?

**Comment by romani** [comment]
@MEZk ,

> Why do we need to follow IDEA Inspection if it is not appropriate in the current situation?

There is design principle that demands to treat all classes around as enemies. No friends at all. Class should not be damaged even around unreasonable hell is going on. 
Immutable design is partly based on this principle.

IntelijIdea do make sense. If you want to ignore it, you need to keep only one ArrayList and share reference to it in all application (as we do know that nobody will change it).
You propose to have 2 copies, from 2 to 3 - there is no big difference.
We already kind of fanatically follow all static analysis rules, .... for example demanding final on variables, there are more and more cases. All of them do have their reason even minor , but reason.  

> So, do you support here our idea to create a new list each time the file is opened?

there is no other ways , how can I be against.

>You mean in getter? Why?

in `getLinesToPosition` as we need to be lazy. If you stand from position that you do not know how class is used, no extra copy should be there if nobody ask for it.

>in destination (CodeSelectorPresentation)

because from perspective of CodeSelectorPresentation there is no guaranty that list is not going to be modified, after it is set in CodeSelectorPresentation. Complete protection, a bit fanatical but .... .
If we start to use special type of immutable collection, then we have clear excuse to suppress - collection is immutable, content in immutable so it it is ok to share references. But we use mutable type so it is design problem and you have to pay for it.
Unfortunately there is no immutable collections in jdk, there bunch of them in other libraries, but extra dependencies are ok only for good reason.

-----------------

We could use https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/ImmutableList.html (we try to minimize guava, but heavily depend on it, so extra change in import control is ok, we use from guava collections that are not present in jdk for now, so it will just another). 
So we make immutable list ones at MainFrameModel and without any copy share it others. 
I still think that refactoring of TreeTable make sense.

**Comment by romani** [comment]
@MEZk , @Kietzmann , @rnveach 

lets vote:
1) ImmutableList (if any complains from IntelijIdea - use suppressions)
2) keep only 3 copies

I am in favor of "1)".

**Comment by rnveach** [comment]
> we try to minimize guava, but heavily depend on it

Guava is issue in code where other utilities like sonar try to incorporate with us. IMO, GUI shouldn't be part of restriction as no one will try to hook into it, so I think we should allow Guava there. If Guava is removed else where and it is an issue having the dependency for GUI, we could separate GUI to a separate JAR.

> lets vote:

I am fine with point 1.

**Comment by Kietzmann** [comment]
I'm fine with 1st point.

**Comment by MEZk** [comment]
@romani 
> There is design principle that demands to treat all classes around as enemies. No friends at all. Class should not be damaged even around unreasonable hell is going on around.
Immutable design is partly based on this principle.

Yes, I know this. It is not a point of my question. Why cannot we create the unmodifiable list view immediately after array list instantiation:
1. Open file
2. Create the new array list 
3. Get unmodifiable copy of the array list
4. Set the copy as the field value.

> IntelijIdea do make sense. If you want to ignore it, you need to keep only one ArrayList and share reference to it in all application (as we do know that nobody will change it).

If somebody try to modify unmodifiable list he will get the exception as the list does not support operations which can change it. The idea of fail fast will work here. We prevent the error from spreading through the system to other classes.

> in getLinesToPosition as we need to be lazy. If you stand from position that you do not know how class is used, no extra copy should be there if nobody ask for it.

Thus, new unmodifiable view will be created each time somebody access the getter, isn't it? In addition, unmodifiable list from Java collection framework is a lightweight class, but immutable collection from Guava is not. Creation of a new unmodifiable list in getter will damage the performance a lot. 

> we try to minimize guava, but heavily depend on it, so extra change in import control is ok, we use from guava collections that are not present in jdk for now, so it will just another).

As long as we continue using Guava and adding new dependencies on it, we will increase the desire to use the library on and on.  May be it is a design problem, if we cannot avoid usage of extra library?

> lets vote:

I'm agains the option number 1. I still cannot understand why we need so many copies. What is the problem with the actions described in my comment https://github.com/checkstyle/checkstyle/pull/4306#issuecomment-301253639 ?

**Comment by romani** [comment]
@MEZk , 

You can make unmodifiable list right at place where we create this collections. Do who get to know that collection is not for change ? It is hidden feature, no special type that you can see on code. You will get to know this only in runtime, not a compilation time. So usage if special type is preferable. Performance is not a point there at all, user will not see any difference.

>May be it is a design problem, if we cannot avoid usage of extra library?

Yes, this is known design problem of jdk, their collections stuck in 2000 year but word go further. We already use guava collections only by this reason.

>What is the problem with the actions described in my comment #4306 (comment) ? 

It is not clear that collection is unmodifiable from simple reading a code (surprise in runtime is not ok), IntellijIdea is right and it force us to think about better approaches.

**Comment by MEZk** [comment]
@romani 
> It is hidden feature, no special type that you can see on code. You will get to know this only in runtime, not a compilation time.

The classes-enemies outside  should not modify the state of CodeSelectorPresentation. If somebody can do this it is a design mistake which violates one of the main OOP principles. 

> It is not clear that collection is unmodifiable from simple reading a code (surprise in runtime is not ok)

CodeSelectorPresentation need to be the only class who knows how to change the array list. Other classes should only read the list. By following the IDEA Inspections we try to cover our design mistakes with extra dependency.

**Comment by romani** [comment]
@Kietzmann , please proceed with  Immutablelist, I persuaded mezk in private chat. Please use Immutablelist as return type from methods and arguments to clearly define that collection is not for changes

**Comment by Kietzmann** [comment]
@romani 
Ok

**Comment by Kietzmann** [comment]
@romani @MEZk 
`linesToPosition` field in `MainFrameModel.class` should be left with no type modification? Am I right?

**Comment by MEZk** [comment]
Summary of our discussion with @romani:

1. Method ```getLinesToPosition``` should return ImmutableList from Guava to make it clear for the client of the class API that he/she works with immutable collection;
2. The array list need to be created in ```openFile``` and wrapped into ImmutableLis to avoid the new immutable list creation each time the client of MainFrameModel invokes ```getLinesToPosition```;
3. ```linesToPosition``` should be ImmutableList;
4. final modifier should be removed from the field ```linesToPosition```. Default value of ```linesToPosition``` is an empty immutable list;
5. If there are violations from IDEA Inspections, Checkstyle, or PMD, they should be suppressed with the description that we use abstract class instead of interface intentionally;
6. Ussage of ImmutableList need to be allowed for MainFrameModel in import_control.xml.

@romani 
Correct me, in case I missed something.

@rnveach 
Please, share your concerns, I did not understand the arguments you gave in our private discussion clearly, sorry.

**Comment by rnveach** [comment]
> Please, share your concerns

I have no concerns, I think I confused myself about part of the GUI code.

**Comment by MEZk** [comment]
@rnveach 
Thanks for the answer.

@Kietzmann 
Please, make changes as described in my comment.

**Comment by romani** [comment]
@MEZk , all is correct.

**Comment by romani** [comment]
@Kietzmann , do you have problems to resolve CI issues ?

**Comment by MEZk** [comment]
@Kietzmann 
Ping

**Comment by Kietzmann** [comment]
I'm here
Currently busy with university assignments(

2017-05-30 20:18 GMT+03:00 Andrei Selkin <notifications@github.com>:

**Comment by rnveach** [comment]
@Kietzmann ping
Please let us know when you plan to continue, otherwise we can close the PR and let someone else finish if they want.

**Comment by Kietzmann** [comment]
@rnveach 
I made required changes recently and need to fix TC build fail

**Comment by romani** [comment]
@Kietzmann , let us know if need help with inspections suppressions.

**Comment by romani** [comment]
Please rebase on latest master, we did travis fix, it should pass of reason was "goodbye" error.

**Comment by romani** [comment]
checkstyle 8.0 was released, please rebase on latest master. 
after each release it is mandatory to rebase to avoid CIs failures.

**Comment by romani** [comment]
@Kietzmann , ping, lets finish this PR.
If you do not have time to continue please let us know, smb else could continue your work.

**Comment by Kietzmann** [comment]
@romani
I'm on it

**Comment by Kietzmann** [comment]
@romani 
I've merged with latest master, looking for all CIs pass.

**Comment by rnveach** [comment]
@Kietzmann TC is still failing.
Did you apply the fixes we asked for at https://github.com/checkstyle/checkstyle/pull/4306#issuecomment-303793608 to make it pass?

**Comment by Kietzmann** [comment]
@rnveach
Could you please give some hint how to suppress IDEA Inspections? Thanks.

**Comment by rnveach** [comment]
@Kietzmann We are expecting the violations to be fixed, not suppressed. See my previous post and the link to the final discussion on what changes we expect to see.

**Comment by Kietzmann** [comment]
@rnveach 
1-4 and 6 are done

**Comment by rnveach** [comment]
> MainFrameModel.java (1)
223: openFile() this is unnecessary in this context

This shouldn't be suppressed. I am looking over the others to verify.

**Comment by rnveach** [comment]
> 49: CodeSelectorPresentation() Assignment to Collection field this.lines2position from parameter 'lines2position'
59: CodeSelectorPresentation() Assignment to Collection field this.lines2position from parameter 'lines2position'
235: setLinePositionMap() Assignment to Collection field this.linePositionMap from parameter 'linePositionMap'
181: getLinesToPosition() 'return' of Collection field linesToPosition 


These look like they are ok to be suppressed based on conversation.
`@noinspection` in javadoc is used to suppress TeamCity. See https://github.com/checkstyle/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/AbstractNestedDepthCheck.java#L30 as an example.

**Comment by Kietzmann** [comment]
@rnveach 
Thanks a lot

**Comment by Kietzmann** [comment]
@rnveach 
Looks like it is done.

**Comment by rnveach** [comment]
@Kietzmann Please give commit message a more descriptive message as this isn't an issue on `Suppressed warnings`. Original message `refactoring CodeSelectorPModel do double copy of collection in c-tor` was better.

File: src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java [hunk 1]
```diff
@@ -232,10 +230,10 @@ public void setEditor(JTextArea textArea) {
     /**
      * Sets line position map.
      * @param linePositionMap Line position map.
+     * @noinspection AssignmentToCollectionOrArrayFieldFromParameter
      */
     public void setLinePositionMap(List<Integer> linePositionMap) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java]: argument type should ImmutableList, in other case it is not clear why we suppress violation

File: src/main/java/com/puppycrawl/tools/checkstyle/gui/CodeSelectorPresentation.java [hunk 2]
```diff
@@ -45,22 +43,22 @@
...
+        this.lines2position = lines2position;
     }
 
     /**
      * Constructor.
      * @param node DetailNode node.
      * @param lines2position list to map lines.
+     * @noinspection AssignmentToCollectionOrArrayFieldFromParameter
      */
     public CodeSelectorPresentation(DetailNode node, List<Integer> lines2position) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/gui/CodeSelectorPresentation.java]: argument type should ImmutableList, in other case it is not clear why we suppress violation

File: src/main/java/com/puppycrawl/tools/checkstyle/gui/CodeSelectorPresentation.java [hunk 3]
```diff
@@ -45,22 +43,22 @@
      * Constructor.
      * @param ast ast node.
      * @param lines2position list to map lines.
+     * @noinspection AssignmentToCollectionOrArrayFieldFromParameter
      */
     public CodeSelectorPresentation(DetailAST ast, List<Integer> lines2position) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/gui/CodeSelectorPresentation.java]: argument type should ImmutableList, in other case it is not clear why we suppress violation

File: src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java (hunk 1 above)
Comment (Kietzmann) [inline src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/gui/CodeSelectorPresentation.java (hunk 3 above)
Comment (Kietzmann) [inline src/main/java/com/puppycrawl/tools/checkstyle/gui/CodeSelectorPresentation.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/gui/CodeSelectorPresentation.java (hunk 2 above)
Comment (Kietzmann) [inline src/main/java/com/puppycrawl/tools/checkstyle/gui/CodeSelectorPresentation.java]: Done

**Pull request merged**